icu/icu4c/source/common/uprefs.cpp

```diff
@@ -0,0 +1,553 @@
+
+        default:
+            return CharString();
+    }
+}
+
+// Maps from a NLS Measurement system to a BCP47 U extension measurement system.
+// 
+// NLS defines:
+// 0 -> Metric system, 1 -> U.S. System
+//
+// This does not return a full nor valid BCP47Tag, it only returns the option that the BCP47 tag
+// would return after the "ms-" part
+//
+// For example:
+//   0 (Metric) would return "metric".
+//   6 (U.S. System) would return "ussystem".
+// 
+// These could be used in a BCP47 tag like this: "en-US-u-ms-metric".
+CharString getMeasureSystemBCP47FromNLSType(int32_t measureSystem, UErrorCode *status) 
+{
+    switch(measureSystem)
+    {
+        case 0:
+            return CharString("metric", *status);
+        case 1:
+            return CharString("ussystem", *status);
+        default:
+            return CharString();
+    }
+}
+
+// -------------------------------------------------------
+// --------------- END OF MAPPING FUNCTIONS --------------
+// -------------------------------------------------------
+
+// -------------------------------------------------------
+// ------------------ HELPER FUCTIONS  -------------------
+// -------------------------------------------------------
+
+// Return the CLDR "h12" or "h23" format for the 12 or 24 hour clock.
+// NLS only gives us a "time format" of a form similar to "h:mm:ss tt"
+// The NLS "h" is 12 hour, and "H" is 24 hour, so we'll scan for the
+// first h or H.
+// Note that the NLS string could have sections escaped with single
+// quotes, so be sure to skip those parts. Eg: "'Hours:' h:mm:ss"
+// would skip the "H" in 'Hours' and use the h in the actual pattern.
+CharString get12_or_24hourFormat(wchar_t* hourFormat, UErrorCode* status)
+{
+    bool isInEscapedString = false;
+    const int32_t hourLength = static_cast<int32_t>(uprv_wcslen(hourFormat));
+    for (int32_t i = 0; i < hourLength; i++)
+    {
+        // Toggle escaped flag if in ' quoted portion
+        if (hourFormat[i] == L'\'') 
+        {
+            isInEscapedString = !isInEscapedString;
+        }
+
+        if (!isInEscapedString) 
+        {
+            // Check for both so we can escape early
+            if (hourFormat[i] == L'H') 
+            {
+                return CharString("h23", *status);
+            }
+
+            if (hourFormat[i] == L'h')
+            {
+                return CharString("h12", *status);
+            }
+        }
+    }
+    // default to a 24 hour clock as that's more common worldwide
+    return CharString("h23", *status);
+}
+
+UErrorCode getUErrorCodeFromLastError()
+{
+    DWORD error = GetLastError();
+    switch(error)
+    {
+        case ERROR_INSUFFICIENT_BUFFER:
+            return U_BUFFER_OVERFLOW_ERROR;
+
+        case ERROR_INVALID_FLAGS:
+        case ERROR_INVALID_PARAMETER:
+            return U_ILLEGAL_ARGUMENT_ERROR;
+        case ERROR_OUTOFMEMORY:
+            return U_MEMORY_ALLOCATION_ERROR;
+        default:    
```

Comment:
super nit: trailing whitespace

```suggestion
        default:
```